src/jace/translator/jaxpr_translator_builder.py

```diff
@@ -0,0 +1,828 @@
+
+        if self.is_root_translator():
+            # The translation, as a whole has finished, so restore the builder,
+            #  i.e. delete all the shared state.
+            self._jax_name_map = {}
+
+        # Remove the current head stack.
+        return self._ctx_stack.pop()
+
+    def _translate_single_eqn(
+        self,
+        eqn: jax_core.JaxprEqn,
+    ) -> tuple[Sequence[str | None], Sequence[str]]:
+        """Translate `eqn` into its SDFG equivalent.
+
+        To do this the function will perform the following steps:
+        - Assemble the in and output variables.
+        - Select the appropriate primitive translator to use.
+        - Create a new empty state terminal state.
+        - Call the primitive translator to perform the translation inside the new state.
+
+        Returns:
+            The SDFG names that were used as inputs and outputs. The inputs might contain `None`
+            which indicates that this particular input was a literal.
+        """
+        if len(eqn.effects) != 0:
+            raise NotImplementedError(f"Equation '{eqn}' has side effects.")
+
+        # Input/Output variables
+        #  Using a tuple for the input ensures that it cannot be modified.
+        in_var_names: Sequence[str | None] = tuple(
+            self.create_jax_var_list(
+                eqn.invars,
+                prevent_creation=True,  # Inputs must already exists.
+                handle_literals=True,  #   but they can be literals.
+            )
+        )
+        out_var_names: MutableSequence[str] = self.create_jax_var_list(
+            eqn.outvars,
+            only_creation=True,  # Output must not exist yet.
+            update_var_mapping=True,
+        )
+
+        pname: str = eqn.primitive.name
+        if pname not in self._primitive_translators:
+            raise NotImplementedError(f"No translator known to handle '{pname}'.")
+        ptranslator = self._primitive_translators[pname]
+
+        # Create the state into which the equation should be translated
+        eqn_state = self.append_new_state(
+            label=f"{pname}_{'_'.join(out_var_names)}",
+            prev_state=None,  # forces the creation of a new terminal state
+        )
+
+        # Now perform the actual translation of the equation.
+        new_sdfg_term_state = ptranslator(
+            builder=self,
+            in_var_names=in_var_names,
+            out_var_names=out_var_names,  # Might be modified by the translator!
+            eqn=eqn,
+            eqn_state=eqn_state,
+        )
+
+        # Determine the new (tentative) terminal state of the SDFG we are building.
+        if new_sdfg_term_state is None:
+            if eqn_state is not self._ctx.terminal_state:
+                raise RuntimeError("Inconsistent terminal state was detected.")
+            new_sdfg_term_state = eqn_state
+        if not self._ctx.validate():
+            raise RuntimeError("Detected an invalid SDFG under construction.")
```

Comment:
Yes, because it essentially checks if the primitive translator mixed up.
You have to keep in mind that this, although it look as if it is very deep inside Jace, is actually user facing code.